src/sagas/dependency.saga.js

```diff
@@ -0,0 +1,325 @@
+// @flow
+
+import { select, call, put, cancel, takeEvery } from 'redux-saga/effects';
+import { getPathForProjectId } from '../reducers/paths.reducer';
+import {
+  loadProjectDependency,
+  loadProjectDependencies,
+} from '../services/read-from-disk.service';
+import {
+  installDependency,
+  uninstallDependency,
+  installDependencies,
+  uninstallDependencies,
+} from '../services/dependencies.service';
+import {
+  ADD_DEPENDENCY_START,
+  ADD_DEPENDENCY_ERROR,
+  ADD_DEPENDENCY_FINISH,
+  UPDATE_DEPENDENCY_START,
+  UPDATE_DEPENDENCY_ERROR,
+  UPDATE_DEPENDENCY_FINISH,
+  DELETE_DEPENDENCY_START,
+  DELETE_DEPENDENCY_ERROR,
+  DELETE_DEPENDENCY_FINISH,
+  ADD_DEPENDENCIES_START,
+  ADD_DEPENDENCIES_ERROR,
+  ADD_DEPENDENCIES_FINISH,
+  DELETE_DEPENDENCIES_START,
+  DELETE_DEPENDENCIES_ERROR,
+  DELETE_DEPENDENCIES_FINISH,
+  addDependencyFinish,
+  addDependencyError,
+  updateDependencyFinish,
+  updateDependencyError,
+  deleteDependencyFinish,
+  deleteDependencyError,
+  addDependenciesStart,
+  addDependenciesError,
+  addDependenciesFinish,
+  deleteDependenciesStart,
+  deleteDependenciesError,
+  deleteDependenciesFinish,
+  showNotification,
+  updateNotification,
+  completeNotification,
+  failNotification,
+} from '../actions';
+
+// Since redux-saga is intended to handle side-effects of actions,
+// all sagas are run *after* all reducers, even though they are
+// incorporated into redux as a middleware. As such, we cannot rely
+// on a store-based check for whether or not a package.json is locked,
+// as any locking actions will lock it in the store before our saga
+// has a chance to check if it is locked or not. As such, we must handle
+// monitoring the lock status within the saga, so we can control the
+// order of checking.
```

Comment:
Right, I forgot about this ordering. It's because technically redux saga isn't a middleware (I mis-spoke earlier), it's a store enhancer, which is an even lower-level thing with which middlewares are built.

But yeah, thinking about it... I'm not sure why I added a reducer to see if the package.json was write-locked. That info should be entirely derivable, by just checking if any dependencies in this project are in the middle of installing/updating/deleting.